### PR TITLE
[build] [vm] Fix flags declaration in dune file.

### PR DESCRIPTION
**Pull request by ejgallego**
Before Dune 2.7, Dune was implicitly adding the `:standard` set of
variables to our CC invocation for `kernel/byterun`.

This set is `-O2 -fno-strict-aliasing -fwrapv -fPIC`, but after
discussion we want to make it only `-fPIC` and control the rest
ourselves with the genearation of `config/dune.c_flags`

### CODE REVIEW

**Comment by SkySkimmer** [comment]
ping @coq/vm-native-maintainers

**Comment by ppedrot** [comment]
That probably means @silene namely because he's the C expert.

**Comment by silene** [comment]
Unfortunately, I am no Dune expert, so I have no clue why this is a bug and what `:standard` means.

**Comment by ejgallego** [comment]
> Unfortunately, I am no Dune expert, so I have no clue why this is a bug and what `:standard` means.

Flags difference:
```
without this PR:

/usr/bin/gcc -O2 -fno-strict-aliasing -fwrapv -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -Wall -Wno-unused -g -O2 -msse2 -mfpmath=sse -g -I /home/egallego/.opam/dev-coq-master/lib/ocaml -o coq_float64.o -c coq_float64.c

with this PR:

/usr/bin/gcc -O2 -fno-strict-aliasing -fwrapv -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -O2 -fno-strict-aliasing -fwrapv -fPIC -Wall -Wno-unused -g -O2 -msse2 -mfpmath=sse -g -I /home/egallego/.opam/dev-coq-master/lib/ocaml -o coq_float64.o -c coq_float64.c

```
So `:standard` as of today is the set of falgs `-fno-strict-aliasing -fwrapv -fPIC`

**Comment by silene** [comment]
If you did not make a typo, it means that the proposed command line contains an awful lot of duplicated options. Thus, even if it is not perfect, I much prefer the current one. So, nak.

**Comment by ejgallego** [comment]
> Thus, even if it is not perfect, I much prefer the current one. So, nak.

Actually if you forget the mess about duplicated options, they are the same; I can cleanup this PR later, sorry the mess, let me rephrase the question this PR was trying to address:

Should we let Dune automatically add `-O2 -fno-strict-aliasing -fwrapv -fPIC` to the set of flags, or should we handle this by hand?

**Comment by silene** [comment]
> Should we let Dune automatically add `-O2 -fno-strict-aliasing -fwrapv -fPIC` to the set of flags, or should we handle this by hand?

It depends. Flag `-fPIC` is forced upon us by OCaml, so we might just as well let Dune handle it (just in case OCaml supports `-fPIE` one day). Flags `-fno-strict-aliasing` and `-fwrapv` do not make sense for Coq and having them just pessimize the generated code. As for `-O2`, we want to have control over it, which is why it is part of `config/dune.c_flags`.

**Comment by ejgallego** [comment]
> It depends. Flag `-fPIC` is forced upon us by OCaml, so we might just as well let Dune handle it (just in case OCaml supports `-fPIE` one day). Flags `-fno-strict-aliasing` and `-fwrapv` do not make sense for Coq and having them just pessimize the generated code. As for `-O2`, we want to have control over it, which is why it is part of `config/dune.c_flags`.

Ok, thanks for the detailed comment, I will add it to the code and implement this solution.

**Comment by SkySkimmer** [comment]
@ejgallego are you still planning to finish this for 8.15?

**Comment by ejgallego** [comment]
> @ejgallego are you still planning to finish this for 8.15?

I think it is likely I postpone for when we bump dune-lang; let me see if I can find a workaround.

[I prefer we don't bump dune-lang for 8.15]

**Comment by ppedrot** [comment]
Doesn't look ready for 8.16.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@ejgallego Do we have an easy way of showing @silene what the flags look like?

**Comment by ejgallego** [comment]
> @ejgallego Do we have an easy way of showing @silene what the flags look like?

The build log displays them:
```
$ grep coq_interp.c _build/log
(cd _build/default/kernel/byterun && /usr/bin/gcc -O2 -fno-strict-aliasing -fwrapv -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC -Wall -Wno-unused -g -O2 -msse2 -mfpmath=sse -g -I /home/egallego/.opam/dev-coq-master/lib/ocaml -o coq_interp.o -c coq_interp.c)
```

File: kernel/byterun/dune [hunk 1]
```diff
@@ -5,7 +5,7 @@
  (foreign_stubs
   (language c)
   (names coq_fix_code coq_float64 coq_memory coq_values coq_interp)
-  (flags :standard (:include %{project_root}/config/dune.c_flags))))
+  (flags -fPIC (:include %{project_root}/config/dune.c_flags))))
```
Comment (silene) [inline kernel/byterun/dune:8]: Why is this `-fPIC` needed?

Comment (ejgallego) [reply kernel/byterun/dune:8]: @silene I added it following your comment in https://github.com/coq/coq/pull/14519#issuecomment-869822279 , maybe I misunderstood what "forced on us" mean.

Comment (silene) [reply kernel/byterun/dune:8]: What I meant is that, as of today, C code meant to be linked with OCaml code needs to be compiled with `-fPIC`. Assuming that Dune knows how to compile C code (does it?), we should never have to pass it explicitly.

Comment (ejgallego) [reply kernel/byterun/dune:8]: Dune knows how to compile C code but have explicitly asked it not to interfere with our choice of flags , `(use_standard_c_and_cxx_flags false)` , so in a sense we can't have our cake and eat it.

I recall the set of flags that dune uses as `:standard` : `-fno-strict-aliasing -fwrapv -fPIC`

Comment (ejgallego) [reply kernel/byterun/dune:8]: Another way to express stuff like this is `(flags :standard \ -fwrapv (:include ...))` , YMMV.

Comment (silene) [reply kernel/byterun/dune:8]: I don't understand what you are saying. In the build log you have shown, option `-fPIC` is passed twice, so there is just no reason to pass it manually. Moreover, `-fwrapv` is being passed too, so Dune's standard flags are being passed, even if you didn't use `:standard`.

Comment (ejgallego) [reply kernel/byterun/dune:8]: Oh yes, sorry, the value for the option is actually wrong; PR fixed. Effective flags are now:
```
/usr/bin/gcc -fPIC -Wall -Wno-unused -g -O2 -msse2 -mfpmath=sse -g -I /home/egallego/.opam/dev-coq-lsp/lib/ocaml -o coq_interp.o -c coq_interp.c
```

Comment (ejgallego) [reply kernel/byterun/dune:8]: ping @silene

**Review by silene — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by ppedrot** [comment]
I think that this PR broke the compilation on the 32-bit CI, see https://gitlab.com/coq/coq/-/jobs/3339312930.

**Comment by JasonGross** [comment]
If you run with `--display=verbose` you should be able to see `ocamlc_cflags` and `ocamlc_cppflags` in `ocaml_config` in `Dune context` which should reveal what cflags dune thinks we ought to be passing

**Comment by Zimmi48** [comment]
What was the reason for merging the PR without running the full CI?
